target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c

```diff
@@ -0,0 +1,461 @@
+
+	pll->hw.init = &hw_init;
+	pll->id = pll_init->id;
+	spin_lock_init(&pll->lock);
+
+	hw_init.name = pll_init->name;
+	hw_init.ops = pll_init->ops;
+
+	if (!pll_init->parent_name) {
+		ret = -EINVAL;
+		goto err_free_pll;
+	}
+	parent_data.fw_name = pll_init->parent_name;
+	hw_init.parent_data = &parent_data;
+	hw_init.num_parents = 1;
+
+	ret = of_clk_hw_register(pll_init->np, &pll->hw);
+	if (ret)
+		goto err_free_pll;
+
+	clk_hw_register_clkdev(&pll->hw, pll_init->name, NULL);
+	return pll;
+
+err_free_pll:
+	kfree(pll);
+
+	return ERR_PTR(ret);
+}
+
+static struct rtl_pll *rtl_pll_find_desc(struct rtl_pll_data *data, unsigned int clk_id)
+{
+	struct rtl_pll *pll;
+	int idx;
+
+	for (idx = 0; idx < PLL_COUNT; ++idx) {
+		pll = data->plls[idx];
+		if (pll && pll->id == clk_id)
+			return pll;
+	}
+
+	return ERR_PTR(-EINVAL);
+}
+
+static struct clk_hw *rtl_pll_of_clk_hw_get(struct of_phandle_args *clkspec, void *priv)
+{
+	struct rtl_pll_data *data = priv;
+	struct rtl_pll *pll;
+	unsigned int clk_id;
+
+	clk_id = clkspec->args[0];
+	pll = rtl_pll_find_desc(data, clk_id);
+	if (IS_ERR(pll)) {
+		pr_info("Invalid PLL clock ID %d specified\n", clk_id);
+		return ERR_CAST(pll);
+	}
+
+	return pll ? &pll->hw : NULL;
+}
+
+void rtl_pll_hw_unregister(struct rtl_pll *pll)
+{
+	clk_hw_unregister(&pll->hw);
+
+	kfree(pll);
+}
+
+static int rtl_pll_clk_register(struct rtl_pll_data *data)
+{
+	int idx, ret;
+
+	for (idx = 0; idx < PLL_COUNT; ++idx) {
+		const struct rtl_pll_info *info = &pll_info[idx];
+		struct rtl_pll_init_data init = {0};
+
+		init.id = info->id;
+		init.name = info->name;
+		init.parent_name = info->parent_name;
+		init.np = data->np;
+		if (rtl_soc_family == RTL8390_FAMILY_ID)
+			init.ops = &rtl839x_pll_ops;
+		else
+			init.ops = &rtl838x_pll_ops;
+
+		data->plls[idx] = rtl_pll_hw_register(&init);
+		if (IS_ERR(data->plls[idx])) {
+			ret = PTR_ERR(data->plls[idx]);
+			pr_err("Couldn't register PLL hw '%s'\n", init.name);
+			goto err_hw_unregister;
+		}
+	}
+
+	ret = of_clk_add_hw_provider(data->np, rtl_pll_of_clk_hw_get, data);
+	if (ret) {
+		pr_err("Couldn't register PLL provider of '%s'\n",
+			of_node_full_name(data->np));
+		goto err_hw_unregister;
+	}
+	
+
+	return 0;
+
+err_hw_unregister:
+	for (--idx; idx >= 0; --idx)
+		rtl_pll_hw_unregister(data->plls[idx]);
+
+	return ret;
+}
+
+static void rtl_clk_init_dt(struct device_node *np, int soc_family)
+{
+	struct rtl_pll_data *data;
+	int ret;
+
+	rtl_soc_family = soc_family;
+
+	data = rtl_pll_create_data(np);
+	if (IS_ERR(data))
+		return;
+
+	ret = rtl_pll_clk_register(data);
+	if (ret)
+		goto err_free_data;
+
+	return;
+
+err_free_data:
+	kfree(data);
+}
+
+static void __init rtl838x_clk_init_dt(struct device_node *np)
+{
+	pr_info("RTL838X clock init name=%s",np->name);
+	rtl_clk_init_dt(np, RTL8380_FAMILY_ID);
+}
+
+static void __init rtl839x_clk_init_dt(struct device_node *np)
+{
+	pr_info("RTL839X clock init");
+	rtl_clk_init_dt(np, RTL8390_FAMILY_ID);	
+}
```

Comment:
I suppose this is the reason we still have the 'mach' specific functions, can't we do this via devicetree probes?

Reply:
rewritten